pkg/server/algorithm_registry.go

```diff
@@ -0,0 +1,63 @@
+package server
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"fmt"
+)
+
+type Algorithm interface {
+	CheckKeyAlgorithm(crypto.PublicKey) bool
+	CheckHashAlgorithm(crypto.Hash) bool
+}
+
+type EcdsaSha256Algorithm struct{}
+
+func (EcdsaSha256Algorithm) CheckKeyAlgorithm(key crypto.PublicKey) bool {
+	_, ok := key.(*ecdsa.PublicKey)
+	return ok
+}
+
+func (EcdsaSha256Algorithm) CheckHashAlgorithm(hash crypto.Hash) bool {
+	return hash == crypto.SHA256
+}
+
+type Ed25519Algorithm struct{}
+
+func (Ed25519Algorithm) CheckKeyAlgorithm(key crypto.PublicKey) bool {
+	_, ok := key.(ed25519.PublicKey)
+	return ok
+}
+
+func (Ed25519Algorithm) CheckHashAlgorithm(hash crypto.Hash) bool {
+	return hash == crypto.Hash(0)
+}
+
+type AlgorithmRegistry struct {
+	permittedAlgorithms []Algorithm
+}
+
+func NewAlgorithmRegistry(algorithmConfig []string) (*AlgorithmRegistry, error) {
+	var permittedAlgorithms []Algorithm
+	for _, algorithm := range algorithmConfig {
+		switch algorithm {
+		case "ecdsa-sha2-256-nistp256":
```

Comment:
TODO: Put these somewhere as constants since we're using them as default value for the CLI flag.

Also, regardless of whether this code gets moved to `sigstore/sigstore`, should this code rely on an enumeration in `sigstore/protobuf-specs` so we have a single source of truth for what combinations are allowed in the ecosystem?

Reply:
As far as I understand, `sigstore/sigstore` doesn't have a dependency to `protobuf-specs` so we'd need to add that.